lib/CPathfinder.cpp

```diff
@@ -455,13 +526,17 @@ CPathsInfo::CPathsInfo( const int3 &Sizes )
 :sizes(Sizes)
 {
 	hero = nullptr;
-	nodes = new CGPathNode**[sizes.x];
+	nodes = new CGPathNode***[sizes.x];
 	for(int i = 0; i < sizes.x; i++)
 	{
-		nodes[i] = new CGPathNode*[sizes.y];
+		nodes[i] = new CGPathNode**[sizes.y];
 		for(int j = 0; j < sizes.y; j++)
 		{
-			nodes[i][j] = new CGPathNode[sizes.z];
+			nodes[i][j] = new CGPathNode*[sizes.z];
+			for (int z = 0; z < sizes.z; z++)
+			{
+				nodes[i][j][z] = new CGPathNode[EPathfindingLayer::NUM_LAYERS];
```

Comment:
@alexvins I'm still not sure if I want to pre allocate nodes for every layer or allocate them dynamically. There is several different ideas that stand behind each approach:

1. If I'm going to make pathfinder reusable so it's not created from scratch each time then we can actually keep only one pathfinder instance in vcmiclient. So it's possible to pre-allocate "CPathsInfo" graph on pathfinder object construction and later on just "copy" it when we need graph, but I unsure if it's reasonable.
2. In same time most of heroes don't use flying and water walking so we can likely save some memory by not allocating nodes for those layers at all.
3. Also pathfinder itself affect how graph allocation/initialization should work. E.g if we check what layers hero can't use for sure on allocation/initialization then we can avoid those checks in pathfinder logic itself.
4. But idea explained in #3 have downside. I like to keep "vcmiserver" as lightweight as possible, but in same time I want to share code that set movement rules. So I want to reuse pathfinder on server for path validation and using graph for validation looks like waste of resources.

So there is plenty of ideas to think about.